L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,504 @@
+  // Sets the certificate provider which is used to store: 
+  // 1. The root certificates used to (cryptographically) verify peer certificate chains.
+  // 2. The certificate chain conveying the application's identity and the corresponding private key.
+  void set_certificate_provider(
+      std::shared_ptr<CertificateProviderInterface> certificate_provider);
+
+  // Watches the updates of root certificates with name |root_cert_name| (set by
+  // the user via set_root_cert_name).  If used in TLS credentials, setting this
+  // field is optional for both the client side and the server side.
+  // If this is not set on the client side, we will use the root certificates
+  // stored in the default system location, since client side must provide root
+  // certificates in TLS (no matter single-side TLS or mutual TLS).
+  // If this is not set on the server side, we will not watch any root
+  // certificate updates, and assume no root certificates needed for the server
+  // (in the one-side TLS scenario, the server is not required to provide root
+  // certificates). We don't support default root certs on server side.
+  void watch_root_certificates();
+
+  // Sets the name of root certificates being watched, if |watch_root_certificates| is
+  // called. If not set, an empty string will be used as the name. For simple
+  // use cases this is not needed, but a certificate provider could be
+  // implemented that watches many sets of certificates. This allows the user to
+  // differentiate between different root certificates.
+  // 
+  // @param root_cert_name the name of root certificates being set.
+  void set_root_cert_name(const std::string& root_cert_name);
+
+  // Watches the updates of identity key-certificate pairs with name
+  // |identity_cert_name|. If used in TLS credentials, it is required to be set
+  // on the server side, and optional for the client side(in the one-side
+  // TLS scenario, the client is not required to provide identity certificates).
+  void watch_identity_key_cert_pairs();
+
+  // Sets the name of identity key-certificate pairs being watched, if
+  // |watch_identity_key_cert_pairs| is called. If not set, an empty string will
+  // be used as the name.
+  //
+  // @param identity_cert_name the name of identity key-certificate pairs being set.
+  void set_identity_cert_name(const std::string& identity_cert_name);
+
+  // WARNING: EXPERT USE ONLY. MISUSE CAN LEAD TO SIGNIFICANT SECURITY DEGRADATION.
+  // 
+  // Sets the TLS session key logging file path. If not set, TLS
+  // session key logging is disabled. Note that this should be used only for
+  // debugging purposes. It should never be used in a production environment
+  // due to security concerns - anyone who can obtain the (logged) session key
+  // can decrypt all traffic on a connection.
+  // 
+  // @param tls_session_key_log_file_path: Path where TLS session keys should
+  // be logged.
+  void set_tls_session_key_log_file_path(
```

Comment:
From review, rename the "expert use only" methods so that the name reflects the danger involved in using this API.

One mild preference was "set_advanced_options(...)" and then isolate them into advanced options.

Reply:
SGTM.

Reply:
I made this change, going to resolve this comment